nodeattr.1
.\"
.\" $Id: nodeattr.1,v 1.6 2003-04-01 22:57:33 achu Exp $
.\" $Source: /g/g0/achu/temp/genders-cvsbackup-full/genders/nodeattr.1,v $
.\"
.\" Copyright (C) 2000-2001 Regents of the University of California
.\" See the DISCLAIMER file distributed with this package.
.\"
.\" Author: Jim Garlick
.\" Adapted from IBM SP version for linux 4/00.
.\"
.TH NODEATTR 1 "3/31/03" "LLNL" "NODEATTR"
.SH NAME
nodeattr \- query genders file
.SH SYNOPSIS
.B nodeattr
.I "[-f genders] [-q | -c | -n | -s] [-r] attr[=val]
.br
.B nodeattr
.I "[-f genders] [-v] [node] attr[=val]
.br
.B nodeattr
.I "[-f genders] -l [node]"
.br
.B nodeattr
.I "[-f genders] -C [node]"
.br
.B nodeattr
.I "[-f genders] --check"
.br
.SH DESCRIPTION
When invoked with the 
.I "-q"
,
.I "-c"
,
.I "-n"
, or 
.I "-s"
arguments,
.B nodeattr
reads the genders file and outputs a list of nodes that have the specified attribute.  
The nodes are listed in hostlist format, comma separated lists, newline separated lists,
or space separated lists.
When called with the 
.I "-r"
option, nodes are listed with their alternate names.  
.LP
When called with a node name (optional) and attribute name,
.B nodeattr
returns \fI0\fR to the environment if the node has the attribute;
else \fI1\fR.
If \fI-v\fR is present, the attribute name and any value (see below) is 
printed on the standard output.  If a node name is not specified, the local 
host is assumed.
.LP
The 
.I "-l"
option causes
.B nodeattr
to print the attributes of the specified node.  If no node is listed, all
attributes in the genders file are listed.
.LP
The
.I "-C"
option causes 
.B nodeattr
to print the local cluster name of a node, or 
.I unknown
if it cannot be determined.  If no node is listed, the local host is assumed.
.LP
The
.I "--check"
option checks the genders file for proper formatting.  Information about
improper formatting will be output to standard error.
.LP
.B Nodeattr 
will always check the default genders file, but a different
genders file is specified with the 
.I "-f"
option.  
.LP
Attribute names may optionally appear in the genders file with an equal sign 
followed by a value.
.B Nodeattr
ignores these values except when the 
.I -v 
option requests that the value, if any, be displayed;  and when an
attribute is specified on the command line with a value, in which case
only an attribute with the specified value in the genders file will match.
.SH EXAMPLES
.LP
To use nodeattr with dsh to run a command on all fddi nodes:
.IP
pdsh -w`nodeattr -c fddi` command
.LP
To use nodeattr in a ksh script to collect a list of users on login nodes:
.IP
for i in `nodeattr -n login`; do rsh $i who; done
.LP
To verify whether or not this node is a head node:
.IP
nodeattr head && echo yes
.LP
.SH "FILES"
/etc/genders
.br
/usr/lib/genders/gendlib.pl
.SH "SEE ALSO"
gendlib(3), libgenders(3)


NEWS
This file describes changes in recent versions of genders. It
primarily documents those changes that are of interest to users
and admins.

Changes in genders-1.4-1
===========================
 -- Added internal indexing and indexing API functions to increase
    search speed for genders.
 -- Added genders querying ability.
 -- Removed maximum attribute and value limitations.

Changes in genders-1.3-1
===========================
 -- Added fixes for the following corner cases:
    - duplicate attributes
    - whitespace between attributes and inline comments
    - trailing whitespace
    - no nodes listed
    - no attributes listed
 -- Added hostrange nodename support to specify ranges of hosts
    rather than a single node on each line of the genders file
 -- Nodes can now be specified on multiple lines
 -- Re-ordered error return code priority within many API functions.
    For example, when an array of length 0 is specified,
    GENDERS_ERR_OVERFLOW may be returned instead of GENDERS_ERR_PARAMETERS.
 
Changes in genders-1.2-1
===========================
 -- Removed implicit "all" attribute from compat/gendlib.pl.  If
    future use of the "all" attribute is required, it must be listed
    as an attribute of every node in the genders database.  See
    the genders.sample file for an example.
 -- Removed use of the /etc/clusters file from compat/gendlib.pl.  If
    future use of a cluster name is required, it must be listed as the
    value of the "cluster" attribute for each node in the genders
    database.  See the genders.sample file for an example.
 -- Removed -r and -C options from nodeattr.  If future use of the
    cluster name is required, the value of the "cluster" attribute can
    be retrieved using the -v option.

Changes in genders-1.2-pre1
===========================
 -- Hostlist code in hostlist.pl (genders-compat package only)
    can now expand ranges without a prefix (.e.g "[0-10]" and "0-10")
    as well as compress back to this same form.
 -- started NEWS file.

$Id: NEWS,v 1.16 2005-01-20 00:54:42 achu Exp $


src/nodeattr/nodeattr.c
/*
 *  * $Id: nodeattr.c,v 1.14 2003-05-08 17:03:24 achu Exp $
 *  * $Source: /g/g0/achu/temp/genders-cvsbackup-full/genders/src/nodeattr/nodeattr.c,v $
 *    
 */

#include <stdio.h>
#include <getopt.h>
#include <assert.h>
#include <string.h>

#include "genders.h"
#include "hostlist.h"

#define HAVE_GETOPT_LONG 1

#if HAVE_GETOPT_LONG
#define GETOPT(ac,av,opt,lopt) getopt_long(ac,av,opt,lopt,NULL)
#else
#define GETOPT(ac,av,opt,lopt) getopt(ac,av,opt)
#endif

#define OPTIONS "cnsqvlCrf:"
static struct option longopts[] = {
    { "querycomma", 0, 0, 'c' },
    { "querynl", 0, 0, 'n' },
    { "queryspace", 0, 0, 'n' },
    { "query", 0, 0, 'q' },
    { "value", 0, 0, 'v' },
    { "cluster", 0, 0, 'C' },
    { "listattr", 0, 0, 'l' },
    { "altnames", 0, 0, 'r' },
    { "filename", 1, 0, 'f' },
    { "check", 0, 0, 'k'},
    { 0,0,0,0 },
};

typedef enum { FMT_COMMA, FMT_NL, FMT_SPACE, FMT_HOSTLIST } fmt_t;

static int test_attr(genders_t gp, char *node, char *attr, int vopt);
static void list_nodes(genders_t gp, char *attr, fmt_t fmt, int ropt);
static void list_attrs(genders_t gp, char *node);
static void usage(void);

/* Utility functions */
static int _gend_error_exit(genders_t gp, char *msg);
static void *_safe_malloc(size_t size);
static void *_rangestr(hostlist_t hl, fmt_t fmt);
static char *_val_create(genders_t gp);
static char *_to_altname(genders_t gp, char *node);
#if 0
static char *_to_gendname(genders_t gp, char *val);
static char *_node_create(genders_t gp);
static char *_attr_create(genders_t gp);
#endif

int
main(int argc, char *argv[])
{
    int c, errors;
    int ropt = 0, lopt = 0, qopt = 0, Copt = 0, vopt = 0, kopt = 0;
    char *filename = GENDERS_DEFAULT_FILE;
    fmt_t qfmt = FMT_HOSTLIST;
    genders_t gp;

    while ((c = GETOPT(argc, argv, OPTIONS, longopts)) != EOF) {
        switch (c) {
        case 'c':   /* --querycomma */
            qfmt = FMT_COMMA;
            qopt = 1;
            break;
        case 'n':   /* --querynl */
            qfmt = FMT_NL;
            qopt = 1;
            break;
        case 's':   /* --queryspace */
            qfmt = FMT_SPACE;
            qopt = 1;
            break;
        case 'q':   /* --query */
            qfmt = FMT_HOSTLIST;
            qopt = 1;
            break;
        case 'v':   /* --value */
            vopt = 1;
            break;
        case 'l':   /* --listattr */
            lopt = 1;
            break;
        case 'C':   /* --cluster */
            Copt = 1;
            break;
        case 'r':   /* --altnames */
            ropt = 1;
            break;
        case 'f':   /* --filename */
            filename = optarg;
            break;
        case 'k':   /* --check */ 
            kopt = 1;
            break;
        default:
            usage();
            break;
        }
    }

    if (optind == argc && !Copt && !lopt && !kopt)
        usage();

    /* Initialize genders package. */
    gp = genders_handle_create();
    if (!gp) {
        fprintf(stderr, "nodeattr: out of memory\n");
        exit(1);
    }

    /* parse check */
    if (kopt) {
        if ((errors = genders_parse(gp, filename, NULL)) == -1) {
          _gend_error_exit(gp, "genders_parse");
        }
        fprintf(stderr, "nodeattr: %d parse errors discovered\n", errors);
        exit(errors);
    }

    if (genders_load_data(gp, filename) < 0)
        _gend_error_exit(gp, filename);

    /* Usage 1: list nodes with specified attribute. */
    if (qopt) {
        char *attr;

        if (vopt || Copt || lopt)
            usage();
        if (optind != argc - 1)
            usage();

        attr = argv[optind++];
        list_nodes(gp, attr, qfmt, ropt);

        exit(0);
    }

    /* Usage 2:  does node have attribute? */
    if (!Copt && !lopt) {
        char *node = NULL, *attr = NULL;
        int result;

        if (optind == argc - 2) {
            node = argv[optind++];
            attr = argv[optind++];
        } else if (optind == argc - 1) {
            node = NULL;
            attr = argv[optind++];
        } else
            usage();

        result = test_attr(gp, node, attr, vopt);
        exit(result ? 0 : 1);
    }

    /* Usage 3:  list attributes */
    if (lopt) {
        char *node = NULL;

        if (optind == argc - 1)
            node = argv[optind++];
        else if (optind != argc)
            usage();

        list_attrs(gp, node);
    }

    /* Usage 4:  list cluster name */
    if (Copt) {
        char *node = NULL;

        if (optind == argc - 1)
            node = argv[optind++];
        else if (optind != argc)
            usage();

        test_attr(gp, node, "cluster", 1);
        exit(0);
    }

    /*NOTREACHED*/
    exit(0);
}

static void list_nodes(genders_t gp, char *attr, fmt_t qfmt, int ropt)
{
    char **nodes;
    int i, count;
    int len = genders_nodelist_create(gp, &nodes);
    hostlist_t hl;
    char *val;
    char *str;

    if ((val = strchr(attr, '=')))  /* attr can actually be "attr=val" */
        *val++ ='\0';
    if ((count = genders_getnodes(gp, nodes, len, attr, val)) < 0)
        _gend_error_exit(gp, attr);

    /* Create a hostlist containing the list of nodes returned by the query */
    hl = hostlist_create(NULL);
    if (hl == NULL) {
        fprintf(stderr, "nodeattr: hostlist_create failed\n");
        exit(1);
    }
    for (i = 0; i < count; i++) {
        char *node = nodes[i];
        char *altnode = NULL;

        /* The -r option just expresses a preference for alternate names, so
         * if the lookup fails, use the primary name.  Making this a hard
         * failure breaks older versions of pdsh that call nodeattr rather
         * than making direct use of libgenders.  -jg
         */
        if (ropt) {
            altnode = _to_altname(gp, nodes[i]);
            if (strlen(altnode) == 0) {
                free(altnode);
                altnode = NULL;
            }
        } 
        if (hostlist_push(hl, altnode ? altnode : node) == 0) {
            fprintf(stderr, "nodeattr: hostlist_push failed\n");
            exit(1);
        }
        if (altnode != NULL)
            free(altnode);
    }
    genders_nodelist_destroy(gp, nodes);

    str = _rangestr(hl, qfmt);
    if (strlen(str) > 0)
        printf("%s\n", str);
    free(str);
    hostlist_destroy(hl);
}

static int test_attr(genders_t gp, char *node, char *attr, int vopt)
{
    char *val = NULL;
    char *wantval;
    int res;

    if ((wantval = strchr(attr, '=')))  /* attr can actually be "attr=val" */
        *wantval++ ='\0';
   
    if (vopt || wantval)
        val = _val_create(gp); /* full of nulls initially */

    if ((res = genders_testattr(gp, node, attr, val, genders_getmaxvallen(gp) + 1)) < 0)
        _gend_error_exit(gp, "genders_testattr");

    if (vopt) {
        if (strlen(val) > 0)
            printf("%s\n", val);
    }
    if (wantval && strcmp(wantval, val) != 0)
        res = 0;
    if (vopt || wantval)
        free(val);
    return res;
}

static void list_attrs(genders_t gp, char *node)
{
    char **attrs, **vals;
    int len, vlen, count, i;

    len = genders_attrlist_create(gp, &attrs);
    vlen = genders_vallist_create(gp, &vals);
    if (node) {
        if ((count = genders_getattr(gp, attrs, vals, len, node)) < 0)
            _gend_error_exit(gp, "genders_getattr");
    } else {
        if ((count = genders_getattr_all(gp, attrs, len)) < 0)
            _gend_error_exit(gp, "genders_getattr_all");
    }
    for (i = 0; i < count; i++)
        if (node && strlen(vals[i]) > 0)
            printf("%s=%s\n", attrs[i], vals[i]);
        else
            printf("%s\n", attrs[i]);
    genders_attrlist_destroy(gp, attrs);
    genders_vallist_destroy(gp, vals);
}

static void usage(void)
{
    fprintf(stderr,
        "Usage: nodeattr [-f genders] [-q|-c|-n|-s] [-r] attr[=val]\n"
        "or     nodeattr [-f genders] [-v] [node] attr[=val]\n"
        "or     nodeattr [-f genders] -l [node]\n"
        "or     nodeattr [-f genders] -C [node]\n"
        "or     nodeattr [-f genders] --check\n");
    exit(1);
}

/**
 ** Utility functions
 **/

static int _gend_error_exit(genders_t gp, char *msg)
{
    fprintf(stderr, "nodeattr: %s: %s\n", 
        msg, genders_strerror(genders_errnum(gp)));
    exit(1);
}

static void *_safe_malloc(size_t size)
{
    void *obj = malloc(size);

    if (obj == NULL) {
        fprintf(stderr, "nodeattr: out of memory\n");
        exit(1);
    }
    memset(obj, 0, size);
    return obj;
}

/* Create a host range string.  Caller must free result. */
static void *_rangestr(hostlist_t hl, fmt_t qfmt)
{
    int size = 65536;
    char *str = _safe_malloc(size);

    /* FIXME: hostlist functions are supposed to return -1 on truncation.
     * This doesn't seem to be working, so make initial size big enough.
     */
    if (qfmt == FMT_HOSTLIST) {
        while (hostlist_ranged_string(hl, size, str) < 0) {
            free(str); 
            size += size;
            str = (char *)_safe_malloc(size);
        }
    } else {
        char sep = qfmt == FMT_SPACE ? ' ' : qfmt == FMT_COMMA ? ',' : '\n';
        char *p;

        while (hostlist_deranged_string(hl, size, str) < 0) {
            free(str); 
            size += size;
            str = (char *)_safe_malloc(size);
        }
        for (p = str; p != NULL; ) {
            if ((p = strchr(p, ',')))
                *p++ = sep;
        }
    }
    return str;
}

/* Create a value string.  Caller must free result. */
static char *_val_create(genders_t gp)
{
    int maxvallen;
    char *val;

    if ((maxvallen = genders_getmaxvallen(gp)) < 0) 
        _gend_error_exit(gp, "genders_getmaxvallen");
    val = (char *)_safe_malloc(maxvallen + 1);

    return val;
}

/* Convert "altname" to "gendname".  Caller must free result.  */
/* Returns zero length string on lookup failure */
static char *_to_altname(genders_t gp, char *node)
{
    char *val = _val_create(gp); /* full of nulls initially */
   
    if (genders_testattr(gp, node, "altname", val, genders_getmaxvallen(gp) + 1) < 0)
        _gend_error_exit(gp, "genders_testattr");
    return val;
}


#if 0
/* Create a node string.  Caller must free result. */
static char *_node_create(genders_t gp)
{
    int maxnodelen;
    char *node;

    if ((maxnodelen = genders_getmaxnodelen(gp)) < 0) 
        _gend_error_exit(gp, "genders_getmaxnodelen");
    node = (char *)_safe_malloc(maxnodelen + 1);
    return node;
}

/* Create an attribute string.  Caller must free result. */
static char *_attr_create(genders_t gp)
{
    int maxattrlen;
    char *attr;

    if ((maxattrlen = genders_getmaxattrlen(gp)) < 0) 
        _gend_error_exit(gp, "genders_getmaxattrlen");
    attr = (char *)_safe_malloc(maxattrlen + 1);
    return attr;
}

/* Convert "altname" to "gendname". Caller must free result. */
static char *_to_gendname(genders_t gp, char *val)
{
    char **nodes;
    int count;
    int len = genders_nodelist_create(gp, &nodes);
    char *node = NULL;

    if ((count = genders_getnodes(gp, nodes, len, "altname", val)) < 0) {
        genders_nodelist_destroy(gp, nodes);
        _gend_error_exit(gp, val);
    } 
    if (count > 1)
        fprintf(stderr, "nodeattr: altname=%s appears more than once!\n", val);

    if (count == 1) {
        node = _node_create(gp);
        strcpy(node, nodes[0]);
    }
    genders_nodelist_destroy(gp, nodes);

    return node;
}
#endif

/*
 * vi:tabstop=4 shiftwidth=4 expandtab
 */
